share/availability/discovery/discovery.go

```diff
@@ -96,26 +104,35 @@ func (d *Discovery) WithOnPeersUpdate(f OnUpdatedPeers) {
 
 // handlePeersFound receives peers and tries to establish a connection with them.
 // Peer will be added to PeerCache if connection succeeds.
-func (d *Discovery) handlePeerFound(ctx context.Context, topic string, peer peer.AddrInfo) {
+func (d *Discovery) handlePeerFound(ctx context.Context, peer peer.AddrInfo, cancelOngoing context.CancelFunc) {
 	if peer.ID == d.host.ID() || len(peer.Addrs) == 0 || d.set.Contains(peer.ID) {
 		return
 	}
-	err := d.set.TryAdd(peer.ID)
+
+	ctx, cancel := context.WithTimeout(ctx, d.dialTimeout)
+	defer cancel()
+	err := d.connector.Connect(ctx, peer)
+	// we don't want to add backoff when the context is canceled.
 	if err != nil {
-		log.Debug(err)
+		if errors.Is(err, context.Canceled) || errors.Is(err, routing.ErrNotFound) {
+			d.connector.RemoveBackoff(peer.ID)
+		}
 		return
 	}
 
-	err = d.connector.Connect(ctx, peer)
+	if uint(d.set.Size()) >= d.peersLimit {
+		log.Debugw("peer limit reached", "count", d.set.Size(), "peer", peer.ID)
+		cancelOngoing()
```

Comment:
Do you think it is reasonable to cancel establishing other connection if we almost done all work for it by this point? I feel it could be very wasteful approach. Could you consider to calculate amount of peers you need before establish ing connection and do not connect to peers you would not require later?

Reply:
> Do you think it is reasonable to cancel establishing other connection if we almost done all work for it by this point? 

Yes. The connect inside does multiple things, and it's not just Dial. Internally, it also does DHT queries through RoutedHost.

> Could you consider to calculate amount of peers you need before establish ing connection and do not connect to peers you would not require later?

This could work, but it's less responsive. Another idea we explored is batching. You know how many connections you need, and you discover this amount of peers through FindPeers, however, getting N peers through FindPeers may take a while, and peers are discovered one by one with intervals in between. Therefore, with the batching approach, we waste time reaching N discovered peers, when we could already start connecting to discover peers immediately.

Reply:
If we already half way to establish the extra connection, maybe just let it settle and have few extra connection in discovery, since we already done so much work to make it?

Reply:
> However, getting N peers through FindPeers may take a while, and peers are discovered individually with intervals in between. Therefore, with the batching approach, we waste time reaching N discovered peers, when we could have already connected to discover peers immediately. The other problem with batching is that you would need to restart FindPeers if some of the N peers are unreachable

I'm suggesting to initiate FindPeers only when amount_connected + amount_connecting < target_amount. One of the possible solutions to implement it would be to trigger FindPeers in event based manner in another routine.

Reply:
> If we already half way to establish the extra connection, maybe just let it settle and have few extra connection in discovery, since we already done so much work to make it?

Actually, I tend to agree with this. So we would cancel only the FindPeers once we got N connections and leave the in-proc connection untouched 👍🏻. It does not have to have a few more connections.

Reply:
> I'm suggesting to initiate FindPeers only when amount_connected + amount_connecting < target_amount. One of the possible solutions to implement it would be to trigger FindPeers in event based manner in another routine

If we allow only one FindPeers at a time, we wouldn't need to track the amount of connecting.

Reply:
What about introducing a buffer of peers? So, we can connect N(where N=available size of peerSet) amount of peers at a time. All other discovered peers will be added to the buffer and in case if we failed at some point, we will get a peer from the buffer.

Reply:
Does the buffer contain connected/connecting peers?

Reply:
Discovered recently.

Reply:
> If we allow only one FindPeers at a time, we wouldn't need to track the amount of connecting.
Thats is part of my suggestion. Have single routine that is responsible to find enough peers. It could be trigger `FindPeers` by external notification(by channel) when amount_connected + amount_connecting drops below target_amount (could be done from disconnect event subscriber). Essentially communicate the state.

Reply:
> We want to connect immediately so each peer we discover instantly goes into 'connecting' state

I got this. But I'm proposing to connect only peerSet.Size() at a time. All other peers will wait. As we now have 1 minute dealing timeout will not take much time.

Reply:
Ok, so this idea might bet useful in case we want to minimize going beyond the limit with possible tradeoff where we defer connecting.

Reply:
This buffering can be implemented over errgroup as it provides rate limiting API

Reply:
Ok, made it